### PR TITLE
feat: added mwan apis

### PR DESCRIPTION
**Pull request by Tbaile**
Adding creation APIs for MWAN

### CODE REVIEW

**Comment by Tbaile** [comment]
Might need additional help here @gsanchietti 
Right now the API just automates the creation process of every part of a MWAN (`interfaces`, `members`, `policy`), however there's no validation whatsoever, how much we want to validate the data?

**Comment by gsanchietti** [comment]
> how much we want to validate the data?

I'd say we should validate all data that can breaks the software.
We can ignore that data that generates and invalid configuration but do not stop the service.

**Comment by Tbaile** [comment]
The only thing that breaks everything is when there's a configuration with the same name as something else. This is extremely dangerous with rules and policies.
Then the only validation step that will be added is:
- roles and policies must have unique names

File: src/nethsec/mwan/__init__.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        tuple of booleans, first one indicates if mwan interface was created, second one indicates if metric was added
+        to network interface
+
+    Raises:
+        ValidationError: if interface name is not defined in /etc/config/network
+    """
+    # checking if interface is configured
+    available_interfaces = utils.get_all_by_type(e_uci, 'network', 'interface')
+    if name not in available_interfaces.keys():
+        raise ValidationError('name', 'invalid', name)
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:57]: Avoid 2 loops here:
```suggestion
    try:
        e_uci.get('network', name)
    except:
        raise ValidationError('name', 'invalid', name)
```

File: src/nethsec/mwan/__init__.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+    created_interface = False
+    # if no interface with name exists, create one with defaults
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'interface').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:61]: Avoid 2 loops here:
```suggestion
    if not e_uci.get('mwan3', 'name', default=None):
```

File: src/nethsec/mwan/__init__.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    if interface_metrics is None:
+        interface_metrics = list[int]()
+        for interface in utils.get_all_by_type(e_uci, 'network', 'interface').values():
+            if 'metric' in interface:
+                interface_metrics.append(int(interface['metric']))
+
+    if metric not in interface_metrics:
+        return metric
+    else:
+        return __generate_metric(e_uci, interface_metrics, metric + 1)
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:37]: Why using a recursive function when a simple loop is enough?

Something like:
```json
    next_metric = 0
    for interface in utils.get_all_by_type(e_uci, 'network', 'interface').values():
        if 'metric' in interface:
            next_metric = max(next_metric, int(interface['metric']))
    return  next_metric + 1
```

You can also remove extra params interface_metrics and metric

File: src/nethsec/mwan/__init__.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    available_interfaces = utils.get_all_by_type(e_uci, 'network', 'interface')
+    if name not in available_interfaces.keys():
+        raise ValidationError('name', 'invalid', name)
+
+    created_interface = False
+    # if no interface with name exists, create one with defaults
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'interface').keys():
+        created_interface = True
+        # fetch default configuration and set interface
+        default_interface_config = utils.get_all_by_type(e_uci, 'ns-api', 'defaults_mwan').get('defaults_mwan')
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:64]:
```suggestion
        default_interface_config = e_uci.get('ns-api', 'defaults_mwan')
```

File: src/nethsec/mwan/__init__.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        metric: metric of the member
+        weight: weight of the member
+
+    Returns:
+        tuple of string and boolean, first one is the generated name of the member, second one indicates if the member
+        was created
+    """
+    member_config_name = utils.get_id(f'{interface_name}_M{metric}_W{weight}')
+    changed = False
+    if member_config_name not in utils.get_all_by_type(e_uci, 'mwan3', 'member').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:108]: Avoid a loop also here.

File: src/nethsec/mwan/__init__.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+    Raises:
+        ValidationError if name is not unique or policy is not valid
+    """
+    rule_config_name = utils.get_id(name.lower(), 15)
+    rules = utils.get_all_by_type(e_uci, 'mwan3', 'rule').keys()
+    if rule_config_name in e_uci.get('mwan3').keys():
+        raise ValidationError('name', 'unique', name)
+    if policy not in utils.get_all_by_type(e_uci, 'mwan3', 'policy').keys():
+        raise ValidationError('policy', 'invalid', policy)
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:143]: Avoid 2 loops and use `_uci.get("mwan3", <key>)`

File: src/nethsec/mwan/__init__.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+    Raises:
+        ValidationError: if name is not unique
+    """
+    changed_config = []
+    # generate policy name
+    policy_config_name = utils.get_id(name.lower())
+    # make sure name is not something that already exists
+    if policy_config_name in e_uci.get('mwan3').keys():
+        raise ValidationError('name', 'unique', name)
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:183]: Avoid loop

File: src/nethsec/mwan/__init__.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+                members.append(__parse_member(e_uci, member))
+
+        # infer policy type by metrics
+        metrics = [int(member['metric']) for member in members]
+        if all(metric == metrics[0] for metric in metrics):
+            policy_data['type'] = 'balance'
+        elif all(metrics.index(metric) == key for key, metric in enumerate(metrics)):
+            policy_data['type'] = 'backup'
+        else:
+            policy_data['type'] = 'custom'
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:272]: While this looks good, why not adding a `ns_tag` (or similar) to the policy to track it easily?
Just an idea, feel free to ignore it.

File: src/nethsec/mwan/__init__.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+                                                            interface['weight'])
+        member_names.append(member_config_name)
+        if member_created:
+            changed_config.append(f'mwan3.{member_config_name}')
+
+    return member_names
+
+
+def edit_policy(e_uci: EUci, name: str, label: str, interfaces: list[dict]) -> list[str]:
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'policy').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:309]: Avoid the loop

File: src/nethsec/mwan/__init__.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+    e_uci.set('mwan3', name, 'use_member', member_names)
+
+    e_uci.save('mwan3')
+    e_uci.save('network')
+    return changed_config
+
+
+def delete_policy(e_uci: EUci, name: str) -> list[str]:
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'policy').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:326]: Avoid the loop

File: src/nethsec/mwan/__init__.py [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+            'sections': order,
+        })
+    ])
+
+    e_uci.save('mwan3')
+    return order
+
+
+def delete_rule(e_uci: EUci, name: str):
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'rule').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:390]: Avoid the loop

File: src/nethsec/mwan/__init__.py [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+def index_rules(e_uci: EUci) -> list[dict]:
+    data = []
+    rules = utils.get_all_by_type(e_uci, 'mwan3', 'rule')
+    for rule_key in rules.keys():
+        rule_data = {}
+        rule_value = rules[rule_key]
+        rule_data['name'] = rule_key
+        rule_data['policy'] = {}
+        rule_data['policy']['name'] = rule_value['use_policy']
+        if rule_value['use_policy'] in utils.get_all_by_type(e_uci, 'mwan3', 'policy').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:342]: Avoid the loop

**Review by gsanchietti — CHANGES_REQUESTED** [review]

File: src/nethsec/mwan/__init__.py (hunk 8 above)
Comment (Tbaile) [reply src/nethsec/mwan/__init__.py:272]: I'm currently using the name of the policy/routes/metrics `ns_<name_normalized>` to keep track of which policy is created by APIs or not.
Probably I'll add something like it down the line, however it's a simple prefix check on `name` field

File: src/nethsec/mwan/__init__.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+        e_uci.set('mwan3', name, 'timeout', default_interface_config['ping_timeout'])
+        e_uci.set('mwan3', name, 'interval', default_interface_config['ping_interval'])
+        e_uci.set('mwan3', name, 'failure_interval', default_interface_config['ping_failure_interval'])
+        e_uci.set('mwan3', name, 'recovery_interval', default_interface_config['ping_recovery_interval'])
+        e_uci.set('mwan3', name, 'down', default_interface_config['interface_down_threshold'])
+        e_uci.set('mwan3', name, 'up', default_interface_config['interface_up_threshold'])
+
+    added_metric = False
+    # avoid adding metric if already present
+    if 'metric' not in e_uci.get_all('network', name).keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:79]: I missed a loop:
```suggestion
    if  e_uci.get('network', name, 'metric', default=None) is None:
```

File: src/nethsec/mwan/__init__.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+
+    e_uci.delete('mwan3', name)
+    e_uci.save('mwan3')
+    return f'mwan3.{name}'
+
+
+def edit_rule(e_uci: EUci, name: str, policy: str, label: str, protocol: str = None,
+              source_address: str = None, source_port: str = None,
+              destination_address: str = None, destination_port: str = None):
+    if name not in utils.get_all_by_type(e_uci, 'mwan3', 'rule').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:395]: Loop!

File: src/nethsec/mwan/__init__.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+        raise ValidationError('name', 'invalid', name)
+
+    if policy not in utils.get_all_by_type(e_uci, 'mwan3', 'policy').keys():
```
Comment (gsanchietti) [inline src/nethsec/mwan/__init__.py:398]: Loop!

**Review by gsanchietti — CHANGES_REQUESTED** [review]
Almost there!

I do not expect much performance gain, but transforming a call complexity from O(N) to O(1) is always a good thing!

**Review by gsanchietti — APPROVED** [review]

**Pull request merged**